### DISCUSSION TITLE
After postscript installation-->User is created but password couldn't be updated

### DISCUSSION THREAD

**Discussion by tsprasath** [Sunbird-Ed/Community#306] [Installation] [answered]
Hi Team,

Trying to create a user facing the following issue.

`curl --location --request POST 'https://xxxxx/api/user/v1/create' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer XXXXX
--header 'x-authenticated-user-token: XXXXXX
--data-raw '{
  "params": {},
    "request":{                 
      "firstName": "uphrh",
      "lastName": "org admin",
      "channel": "sunbird",
      "userName": "orgadmin1",
      "email": "orgadmin1@yopmail.com",
      "emailVerified": true,
      "password": "Admin@123"
    }
}'`

`{
    "id": "api.user.create",
    "ver": "v1",
    "ts": "2022-12-13 15:09:24:508+0000",
    "params": {
        "resmsgid": "69ebc41408894f00b09f9ad5eb345cae",
        "msgid": "69ebc41408894f00b09f9ad5eb345cae",
        "err": null,
        "status": "SUCCESS",
        "errmsg": null
    },
    "responseCode": "OK",
    "result": {
        "userId": "24a090fa-bb8a-4429-b6f7-aa3ab1b3f341",
        "response": "SUCCESS",
        "err_msg": "User is created but password couldn't be updated.",
        "errors": []
    }
}`

ERROR:

`{"timestamp":"2022-12-13T15:09:24.482Z","msg":"{\"eid\":\"ERROR\",\"ets\":1670944164482,\"ver\":\"3.0\",\"mid\":\"69ebc41408894f00b09f9ad5eb345cae\",\"actor\":{\"id\":\"a97b42ce-7906-414e-bc30-31a894680f53\",\"type\":\"User\"},\"context\":{\"channel\":\"0136855050200760320\",\"pdata\":{\"id\":\"dev.sunbird.learning.service\",\"pid\":\"learner-service\",\"ver\":\"4.10.0\"},\"env\":\"User\",\"cdata\":[{\"id\":\"69ebc41408894f00b09f9ad5eb345cae\",\"type\":\"Request\"}],\"rollup\":{}},\"edata\":{\"err\":\"UOS_0069\",\"stacktrace\":\"updatePassword: Exception occurred:  org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.handleErrorStatus(ClientInvocation.java:216)org.jboss.resteasy.client.jaxrs.internal.proxy.extractors.DefaultEntityExtractorFactory$3.extractEntity(DefaultEntityExtractorFactory.java:50)org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invoke(ClientInvoker.java:107)org.jboss.resteasy.client.jaxrs.internal.proxy.ClientProxy.invoke(ClientProxy.java:76)com.sun.proxy.$Proxy49.resetPassword(Unknown Source)org.sunbird.sso.impl.KeyCloakServiceImpl.updatePassword(KeyCloakServiceImpl.java:75)org.sunbird.util.user.UserUtil.updatePassword(UserUtil.java:298)org.sunbird.service.user.impl.SSOUserServiceImpl.createUserAndPassword(SSOUserServiceImpl.java:99)org.sunbird.actor.user.SSOUserCreateActor.processSSOUser(SSOUserCreateActor.java:117)org.sunbird.actor.user.SSOUserCreateActor.createSSOUser(SSOUserCreateActor.java:89)org.sunbird.actor.user.SSOUserCreateActor.onReceive(SSOUserCreateActor.java:61)org.sunbird.actor.core.BaseActor.onReceive(BaseActor.java:21)akka.actor.UntypedAbstractActor$$anonfun$receive$1.applyOrElse(AbstractActor.scala:291)akka.actor.Actor$class.aroundReceive(Actor.scala:517)akka.actor.AbstractActor.aroundReceive(AbstractActor.scala:180)akka.actor.ActorCell.receiveMessage(ActorCell.scala:588)akka.actor.ActorCell.invoke(ActorCell.scala:557)akka.dispatch.Mailbox.processMailbox(Mailbox.scala:258)akka.dispatch.Mailbox.run(Mailbox.scala:225)akka.dispatch.Mailbox.exec(Mailbox.scala:235)akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)\",\"errtype\":\"api_access\",\"requestid\":\"69ebc41408894f00b09f9ad5eb345cae\"}}","lname":"TelemetryEventLogger","tname":"application-akka.actor.most-used-one-dispatcher-176","level":"INFO","HOSTNAME":"learner-65cd7d54db-6lhvf","application.home":"/home/sunbird/learner/learning-service-1.0-SNAPSHOT"}`

**Comment by tsprasath** [comment, marked as answer]
@keshavprasadms 

Steps followed to troubleshoot.

1. Cassandra provider is available in user federation(Keycloak sunbird realm).
2. Channel are available in cassandra DB.
3. Player service  --> SSO_URL and REALM are set.

**Reply by beepdot** [reply]
@reshmi-nair

**Reply by reshmi-nair** [reply]
@tsprasath Seems like keycloak is not allowing learner service to access the API.
ClientInvocation.java, line number 216 says:
       case 403:
            throw new ForbiddenException(response);
Pls check the keycloak configurations and also check if the keycloak url is accessible from learner pod.

**Reply by tsprasath** [reply]
@keshavprasadms  Problem with Keycloak environment variable 

Ingress IP is not updated properly the values can be visible in keycloak configuration.

solution:

1. Build and redeploy the keycloak from Jenkins.Then check in the keycloak VM by the following command
`cat /etc/environment`
`sunbird_user_service_base_url=http://<INGRESS IP>/learner`
After trying to reset the password and try to update it.
Still, if the issue faces by the captcha. In player service, configmap  -->ensure the below variables are set to false
`sunbird_p1_reCaptcha_enabled: false`
`sunbird_p2_reCaptcha_enabled:false`
`sunbird_p3_reCaptcha_enabled:false`

Then restart the player pod. Started to work.

**Reply by tsprasath** [reply]
@keshavprasadms @reshmi-nair while user creation password is not updated.

`"User is created but password couldn't be updated.", "errors": [] } }`

**Reply by tsprasath** [reply]
Followed below settings in the keycloak

![MicrosoftTeams-image](https://user-images.githubusercontent.com/20708426/208351651-ddc9b89e-3ed3-485a-8af3-0160903133e4.png)

Updated below vars in the private repo with its new values 

-    core_vault_sunbird_sso_publickey 
-    adminutil_refresh_token_public_key_kid 
-    adminutil_refresh_token_secret_key

Then I tried with user create API and register from the portal and started working.

**Reply by beepdot** [reply]
Thanks for the details @tsprasath 
But all these details we have already documented
https://github.com/project-sunbird/sunbird-devops/blob/master/private_repo/ansible/inventory/dev/Core/secrets.yml#L72
https://github.com/project-sunbird/sunbird-devops/blob/master/private_repo/ansible/inventory/dev/Core/common.yml#L7

If you had filled all the mandatory variables before proceeding with deployments, you would not have encountered these issues.

**Reply by beepdot** [reply]
And recaptcha was part of the initial requirements we have mentioned in installation doc.
And these fall under the mandatory list of variables
https://github.com/project-sunbird/sunbird-devops/blob/master/private_repo/ansible/inventory/dev/Core/secrets.yml#L118-L119

Mandatory vars section:
https://github.com/project-sunbird/sunbird-devops/blob/master/private_repo/ansible/inventory/dev/Core/secrets.yml#L2

Optional and default vars section:
https://github.com/project-sunbird/sunbird-devops/blob/master/private_repo/ansible/inventory/dev/Core/secrets.yml#L123
